server/src/main/java/org/elasticsearch/action/support/MappedActionFilters.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.support;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.tasks.Task;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class MappedActionFilters implements ActionFilter {
+
+    private final Map<String, List<MappedActionFilter>> filtersByAction;
+
+    public MappedActionFilters(List<MappedActionFilter> mappedFilters) {
+        Map<String, List<MappedActionFilter>> map = new HashMap<>();
+        for (var filter : mappedFilters) {
+            map.computeIfAbsent(filter.actionName(), k -> new ArrayList<>()).add(filter);
+        }
+        map.replaceAll((k, l) -> List.copyOf(l));
+        this.filtersByAction = Map.copyOf(map);
```

Comment:
Out of curiosity: I have seen this pattern elsewhere in our code base. Is this to use the HashMap as a "builder", and then store an immutable collection? Is it better than `Collections.unmodifiableMap()`?

Reply:
Map.copyOf builds a brand new map which is immutable. Collections.unmodifiableMap returns a delegated map. The latter will incur an additional indirection on every access. Using Map.copyOf is trading off runtime cost for one time startup cost.

Reply:
TIL. Thanks for the info!